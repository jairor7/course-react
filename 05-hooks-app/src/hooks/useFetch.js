import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });
  const isMounted = useRef(true);
  // const isMounted = useRef(1);

  useEffect(() => {
    return () => {
      isMounted.current = false;
      // isMounted.current = 2;
    };
  }, []);

  useEffect(() => {
    setState({ data: null, loading: true, error: null });
    fetch(url)
      .then((resp) => resp.json())
      .then((data) => {
        // setTimeout(() => {
        // if (isMounted.current) {
        // if (isMounted.current === 1) {
        isMounted.current && setState({ data, loading: false, error: null });
        // setState({ data, loading: false, error: null });
        // } else console.log("No se llamo setState");
        // }, 3000);
      });
  }, [url]);

  return state;
};
