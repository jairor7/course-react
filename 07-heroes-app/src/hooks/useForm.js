import { useState } from "react";

export const useForm = (initialFormState = {}) => {
  const [form, setForm] = useState(initialFormState);

  const reset = () => {
    setForm(initialFormState);
  };

  const handleInputChange = (e) => {
    setForm({
      ...form,
      [e.target.name]: e.target.value,
    });
  };

  return [form, handleInputChange, reset];
};
