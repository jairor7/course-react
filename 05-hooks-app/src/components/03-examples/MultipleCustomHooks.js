import React from "react";
import { useFetch } from "../../hooks/useFetch";
import { useCounter } from "../../hooks/useCounter";
import "../../styles/components/02-useEffect/effect.css";

export const MultipleCustomHooks = () => {
  const [counter, increment] = useCounter(1);
  // console.log(counter);
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );
  const { author, quote } = !!data && data[0];
  // console.log(!!data);
  // console.log(author, quote);

  return (
    <div>
      <h1>BreakingBAD Quotes</h1>
      <hr />

      {loading ? (
        <div className="alert alert-info text-center">
          Loading ...
        </div>
      ) : (
        <div>
          <blockquote className="blockquote text-end">
            <p className="mb-1">{quote}</p>
          </blockquote>
          <figcaption className="blockquote-footer text-end">
            {author}
          </figcaption>
        </div>
      )}
      {!loading && (
        <button
          onClick={() => increment()}
          className="btn btn-primary"
        >
          Siguiente
        </button>
      )}
    </div>
  );
};
