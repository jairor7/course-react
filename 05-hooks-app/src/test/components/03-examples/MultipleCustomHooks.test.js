import { shallow } from "enzyme";
import { MultipleCustomHooks } from "../../../components/03-examples/MultipleCustomHooks";
import { useFetch } from "../../../hooks/useFetch";
import { useCounter } from "../../../hooks/useCounter";
jest.mock("../../../hooks/useFetch");
jest.mock("../../../hooks/useCounter");

describe("Pruebas en MultipleCustomHooks", () => {
  test("Debe de mostrarse correctamente", () => {
    useCounter.mockReturnValue([1]);
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de mostrar la informaciÃ³n", () => {
    useCounter.mockReturnValue([1, () => {}]);
    useFetch.mockReturnValue({
      data: [
        { author: "Jairo", quote: "Nunca digas nunca!!" },
      ],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper.find("blockquote").exists()).toBe(true);
    expect(wrapper.find(".mb-1").text()).toBe(
      "Nunca digas nunca!!"
    );
    expect(wrapper.find("figcaption").text()).toBe("Jairo");
    expect(wrapper.find("button").exists()).toBe(true);
  });
});
