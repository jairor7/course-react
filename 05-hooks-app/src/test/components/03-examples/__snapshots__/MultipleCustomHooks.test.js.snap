// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks Debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    BreakingBAD Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading ...
  </div>
</div>
`;
